### ISSUE TITLE
ModuleNotFoundError first thing out of the box?

### ISSUE THREAD

**Issue by dvpierce** [cloud-custodian/tfparse#144]
I updated to Python 3.11, did a pip install of tfparse, and get: "ModuleNotFoundError: No module named 'tfparse._tfparse'"

Having the same issue on Linux and Windows. Should I install an older version or something? Thanks!

**Comment by kapilt** [comment]
per the readme we distribute binaries wheels for linux and mac, but not windows atm. looking over the pypi files (https://pypi.org/project/tfparse/#files)  there shouldn't be any issue on linux, what linux distro?

**Comment by fornof** [comment]
I downloaded the source and I still don't see where tfparse._tfparse is supposed to come from. - can you provide a link the source where it is?  Could it be that **build_cffy.py** ? 

I tried putting in the ffi from there - I have WSL Ubuntu ,  and now I get: 

```
(venv-310) PS C:\projects\antlr> python index.py
Traceback (most recent call last):
  File "C:\projects\antlr\index.py", line 1, in <module>
    from tfparse import load_from_path
  File "C:\projects\antlr\venv-310\lib\python3.9\site-packages\tfparse\__init__.py", line 41, in <module>
    lib = load_lib()
  File "C:\projects\antlr\venv-310\lib\python3.9\site-packages\tfparse\__init__.py", line 38, in load_lib
    return ffi.dlopen(str(libpath))
  File "C:\projects\antlr\venv-310\lib\python3.9\site-packages\cffi\api.py", line 150, in dlopen
    lib, function_cache = _make_ffi_library(self, name, flags)
  File "C:\projects\antlr\venv-310\lib\python3.9\site-packages\cffi\api.py", line 832, in _make_ffi_library
    backendlib = _load_backend_lib(backend, libname, flags)
  File "C:\projects\antlr\venv-310\lib\python3.9\site-packages\cffi\api.py", line 827, in _load_backend_lib
    raise OSError(msg)
OSError: cannot load library 'C:\projects\antlr\venv-310\lib\python3.9\site-packages\tfparse.cp39-mingw_x86_64_ucrt.pyd': error 0x7e.  Additionally, ctypes.util.find_library() did not manage to locate a library called 'C:\\projects\\antlr\\venv-310\\lib\\python3.9\\site-packages\\tfparse.cp39-mingw_x86_64_ucrt.pyd'
```

the pyd is not on my system. perhaps it needs to be built from source?  in virtual environment I run `pip install .` , note gcc and go must be present to get a success.  
 ```
 ....
 Attempting uninstall: tfparse
    Found existing installation: tfparse 0.2.0
    Uninstalling tfparse-0.2.0:
      Successfully uninstalled tfparse-0.2.0
Successfully installed tfparse-0.6.1
```
now when I run index.py I get: 
```
 python index.py
/mnt/c/projects/antlr dirpath
dict_keys(['kubernetes_cluster_role_binding', 'kubernetes_deployment', 'kubernetes_namespace', 'kubernetes_persistent_volume_claim', 'kubernetes_service', 'kubernetes_service_account', 'locals', 'output', 'terraform', 'variable'])
```

here is what the index.python looks like: 
```
from tfparse import load_from_path
import os 
dir_path = os.path.dirname(os.path.realpath(__file__))
print(dir_path, 'dirpath')
parsed = load_from_path(dir_path +'/terraform-kubernetes-portainer')
print(parsed.keys())
```
and here is the main.tf in the load_from_path path: 
https://github.com/jhole89/terraform-kubernetes-portainer/blob/master/main.tf 


pypi is sorely outdated. It looks like packages can be built from source on WSL as demonstrated. 

I wish there was a step by step through guide on how to use this. Instead I find myself looking at tests. 

My end goal is to read into a format that can be modified (such as a dictionary) with types, modify, then write back to the same file without changes. 

I don't see outer level or inner comments being handled - I expect them to be in parsed.keys() , perhaps as linenumber_comment (L36_comment : {'type' comment,  'line_end': 24, 'line_start': 21}}) 

The output seems like it would be harder to modify. Granted it is better than hcl2 that doesn't include details. 

```
dict_keys(['kubernetes_cluster_role_binding', 'kubernetes_deployment', 'kubernetes_namespace', 'kubernetes_persistent_volume_claim', 'kubernetes_service', 'kubernetes_service_account', 'locals', 'output', 'terraform', 'variable'])
[{'__tfmeta': {'filename': 'main.tf', 'label': 'kubernetes_cluster_role_binding', 'line_end': 35, 'line_start': 20, 'path': 'kubernetes_cluster_role_binding.portainer', 'type': 'resource'}, 'id': '8e743d6a-9164-4011-9612-9849de916090', 'metadata': {'__tfmeta': {'filename': 'main.tf', 'line_end': 24, 'line_start': 21}, 'id': '6bf0435f-6784-4c13-8acf-562939b54fd2', 'name': 'portainer-crb-clusteradmin'}, 'role_ref': {'__tfmeta': {'filename': 'main.tf', 'line_end': 29, 'line_start': 25}, 'api_group': 'rbac.authorization.k8s.io', 'id': 'bc073289-56c5-488c-946e-936fdff6e36b', 'kind': 'ClusterRole', 'name': 'cluster-admin'}, 'subject': {'__tfmeta': {'filename': 'main.tf', 'line_end': 34, 'line_start': 30}, 'id': '5368de74-2f3f-45f9-b330-96daa7ae5421', 'kind': 'ServiceAccount', 'name': {'__attribute__': 'kubernetes_service_account.portainer.metadata[0].name', '__name__': 'portainer', '__ref__': '565d83e1-4199-4c4c-87ef-469ee8e2a7d4', '__type__': 'kubernetes_service_account'}, 'namespace': {'__attribute__': 'kubernetes_namespace.portainer.metadata[0].name', '__name__': 'portainer', '__ref__': '84679534-3805-4c63-ab76-fe541983ea9a', '__type__': 'kubernetes_namespace'}}}]
```

I don't think I can use unless it has comments save in there unfortunately :\ 
Not sure your time frame/availability on this, It looks like it uses GO somehow. I could probably put in a comment fix if someone points me in the right direction. 
Alternatively, I'm looking at antlr4 terraform grammar to look into for comments.

**Comment by fornof** [comment]
I downloaded the source and I still don't see where tfparse._tfparse is supposed to come from. - can you provide a link the source where it is?  Could it be that **build_cffy.py** ? 

I tried putting in the ffi from there - I have WSL Ubuntu ,  and now I get: 

```
(venv-310) PS C:\projects\antlr> python index.py
Traceback (most recent call last):
  File "C:\projects\antlr\index.py", line 1, in <module>
    from tfparse import load_from_path
  File "C:\projects\antlr\venv-310\lib\python3.9\site-packages\tfparse\__init__.py", line 41, in <module>
    lib = load_lib()
  File "C:\projects\antlr\venv-310\lib\python3.9\site-packages\tfparse\__init__.py", line 38, in load_lib
    return ffi.dlopen(str(libpath))
  File "C:\projects\antlr\venv-310\lib\python3.9\site-packages\cffi\api.py", line 150, in dlopen
    lib, function_cache = _make_ffi_library(self, name, flags)
  File "C:\projects\antlr\venv-310\lib\python3.9\site-packages\cffi\api.py", line 832, in _make_ffi_library
    backendlib = _load_backend_lib(backend, libname, flags)
  File "C:\projects\antlr\venv-310\lib\python3.9\site-packages\cffi\api.py", line 827, in _load_backend_lib
    raise OSError(msg)
OSError: cannot load library 'C:\projects\antlr\venv-310\lib\python3.9\site-packages\tfparse.cp39-mingw_x86_64_ucrt.pyd': error 0x7e.  Additionally, ctypes.util.find_library() did not manage to locate a library called 'C:\\projects\\antlr\\venv-310\\lib\\python3.9\\site-packages\\tfparse.cp39-mingw_x86_64_ucrt.pyd'
```

the pyd is not on my system. perhaps it needs to be built from source?  in virtual environment I run `pip install .` , note gcc and go must be present to get a success.  
 ```
 ....
 Attempting uninstall: tfparse
    Found existing installation: tfparse 0.2.0
    Uninstalling tfparse-0.2.0:
      Successfully uninstalled tfparse-0.2.0
Successfully installed tfparse-0.6.1
```
now when I run index.py I get: 
```
 python index.py
/mnt/c/projects/antlr dirpath
dict_keys(['kubernetes_cluster_role_binding', 'kubernetes_deployment', 'kubernetes_namespace', 'kubernetes_persistent_volume_claim', 'kubernetes_service', 'kubernetes_service_account', 'locals', 'output', 'terraform', 'variable'])
```

here is what the index.python looks like: 
```
from tfparse import load_from_path
import os 
dir_path = os.path.dirname(os.path.realpath(__file__))
print(dir_path, 'dirpath')
parsed = load_from_path(dir_path +'/terraform-kubernetes-portainer')
print(parsed.keys())
```
and here is the main.tf in the load_from_path path: 
https://github.com/jhole89/terraform-kubernetes-portainer/blob/master/main.tf 


pypi is sorely outdated. It looks like packages can be built from source on WSL as demonstrated. 

I wish there was a step by step through guide on how to use this. Instead I find myself looking at tests. 

My end goal is to read into a format that can be modified (such as a dictionary) with types, modify, then write back to the same file without changes. 

I don't see outer level or inner comments being handled - I expect them to be in parsed.keys() , perhaps as linenumber_comment (L36_comment : {'type' comment,  'line_end': 24, 'line_start': 21}}) 

The output seems like it would be harder to modify. Granted it is better than hcl2 that doesn't include details. 

```
dict_keys(['kubernetes_cluster_role_binding', 'kubernetes_deployment', 'kubernetes_namespace', 'kubernetes_persistent_volume_claim', 'kubernetes_service', 'kubernetes_service_account', 'locals', 'output', 'terraform', 'variable'])
[{'__tfmeta': {'filename': 'main.tf', 'label': 'kubernetes_cluster_role_binding', 'line_end': 35, 'line_start': 20, 'path': 'kubernetes_cluster_role_binding.portainer', 'type': 'resource'}, 'id': '8e743d6a-9164-4011-9612-9849de916090', 'metadata': {'__tfmeta': {'filename': 'main.tf', 'line_end': 24, 'line_start': 21}, 'id': '6bf0435f-6784-4c13-8acf-562939b54fd2', 'name': 'portainer-crb-clusteradmin'}, 'role_ref': {'__tfmeta': {'filename': 'main.tf', 'line_end': 29, 'line_start': 25}, 'api_group': 'rbac.authorization.k8s.io', 'id': 'bc073289-56c5-488c-946e-936fdff6e36b', 'kind': 'ClusterRole', 'name': 'cluster-admin'}, 'subject': {'__tfmeta': {'filename': 'main.tf', 'line_end': 34, 'line_start': 30}, 'id': '5368de74-2f3f-45f9-b330-96daa7ae5421', 'kind': 'ServiceAccount', 'name': {'__attribute__': 'kubernetes_service_account.portainer.metadata[0].name', '__name__': 'portainer', '__ref__': '565d83e1-4199-4c4c-87ef-469ee8e2a7d4', '__type__': 'kubernetes_service_account'}, 'namespace': {'__attribute__': 'kubernetes_namespace.portainer.metadata[0].name', '__name__': 'portainer', '__ref__': '84679534-3805-4c63-ab76-fe541983ea9a', '__type__': 'kubernetes_namespace'}}}]
```

I don't think I can use unless it has comments save in there unfortunately :\ 
Not sure your time frame/availability on this, It looks like it uses GO somehow. I could probably put in a comment fix if someone points me in the right direction. 
Alternatively, I'm looking at antlr4 terraform grammar to look into for comments.

**Comment by kapilt** [comment]
that your python is looking for a `.pyd` implies your on windows not WSL / ubuntu (for linux its `.so` for a shared library). ie that its a windows python, not a linux one. wrt to the source a checkout and running python setup.py develop / install should do all the things re compilation, as you noted you need golang, a c compiler, and python development headers, libraries. again as your using a windows python and os, I can't really help, but would love to see this work, so please do report back what you find. wrt to pypi outdated, they are updated per release automatically, again, we're looking for help on windows builds.

**Comment by kapilt** [comment]
yeah we don't currently look at comments, so if that's important for your use case, then its likely going to require a different library or a workaround. regarding workaround, we do store source offsets for each block, which allow for reading the block or a its prefix as a string from the source to examine for comments.

**Comment by kapilt** [comment]
re windows support at least that's been addressed in #177 wrt to round tripping and comment preserving, that's not really in scope to the project wrt to producing a pure AST style tree. our focus is more on interpreting the underlying hcl with function invocation and variable interpolation and module downloading to evaluate the intended resources.
